source/common/upstream/upstream_impl.cc

```diff
@@ -298,16 +298,22 @@ HostDescriptionImpl::HostDescriptionImpl(
       priority_(priority),
       socket_factory_(resolveTransportSocketFactory(dest_address, metadata_.get())),
       creation_time_(time_source.monotonicTime()) {
-  if (health_check_config.port_value() != 0 && dest_address->type() != Network::Address::Type::Ip) {
+  const auto& port_value = health_check_config.port_value();
+  if (port_value != 0 && dest_address->type() != Network::Address::Type::Ip) {
     // Setting the health check port to non-0 only works for IP-type addresses. Setting the port
     // for a pipe address is a misconfiguration. Throw an exception.
     throw EnvoyException(
         fmt::format("Invalid host configuration: non-zero port for non-IP address"));
   }
-  health_check_address_ =
-      health_check_config.port_value() == 0
-          ? dest_address
-          : Network::Utility::getAddressWithPort(*dest_address, health_check_config.port_value());
+  if (health_check_config.has_address()) {
+    auto address = Network::Address::resolveProtoAddress(health_check_config.address());
+    health_check_address_ =
+        port_value == 0 ? address : Network::Utility::getAddressWithPort(*address, port_value);
+  } else {
+    health_check_address_ = port_value == 0
+                                ? dest_address
+                                : Network::Utility::getAddressWithPort(*dest_address, port_value);
+  }
```

Comment:
I think this could be factored into a utility function and shared with the logical host code above?